### PR TITLE
Proof of Concept for Zama Functionality

### PR DESCRIPTION
**Pull request by vishakh**
This draft PR builds a PoC for evaluating the suitability of [Zama](https://zama.ai/) for Monadic DNA. 

Done so far:
🧬 Load a 23andMe dataset
🔢 Integer-encode the SNPs and genotypes in the dataset
㊙️ Encrypt genotypes as Zama secrets
🔍 Run lookups for various medical conditions for bechmarking
⏺️ Persist encrypted data to disk and read it back for benchmarking
☎️ Refactor into client / server model to understand client-side encryption and general latency

Left for a future PR:
🔑 Figure out the best way to make key management and usage work

### CODE REVIEW

File: zama-poc/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+    return Ok(());
+}
+
+fn encrypt_genotypes_for_zama(processed_data: &HashMap<u64, u8>, client_key: &ClientKey) -> result::Result<HashMap<u64, FheUint8>, std::io::Error> {
+    let mut enc_data = HashMap::new();
+
+    for (&encoded_rsid, &encoded_genotype) in processed_data.iter() {
+        let genotype_encrypted = FheUint8::try_encrypt(encoded_genotype, client_key)
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;  // Convert tfhe::Error to std::io::Error
```
Comment (bhaskarkishore) [inline zama-poc/src/main.rs:57]: Why not encrypt the rsid? Can't claim fully encrypted without it.

File: zama-poc/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    info!("Number of lines to process: {:?}", num_lines);
+
+    let processed_data = process_file(filename, num_lines)?;
+    info!("Lines of processed data: {:?}", processed_data.len());
+    //println!("{:?}", processed_data.);
+
+    let encrypted_genotypes = encrypt_genotypes_for_zama(&processed_data, &client_key)?;
+    info!("Lines of encrypted data: {:?}", encrypted_genotypes.len());
+
+    return Ok(());
```
Comment (bhaskarkishore) [inline zama-poc/src/main.rs:49]: Might be worth benching only the zama parts and not IO or preprocessing.

**Comment by bhaskarkishore** [comment]
Also trying out in general is compressed keys/secrets: https://docs.zama.ai/tfhe-rs/fundamentals/compress

File: zama-poc/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+name = "main"
+path = "src/main.rs"
+
+[[bin]]
+name = "server"
+path = "src/server.rs"
+
+[[bin]]
+name = "client"
+path = "src/client.rs"
```
[the remark this answers is not in the document]
Comment (vishakh) [reply zama-poc/Cargo.toml:32]: Leaving it to a future PR.

File: zama-poc/scripts/create_bin_file.sh [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# Creates a binary file for test_api.sh
+
+# Create a simple text file
+echo "This is a sample dataset" > sample.txt
```
[the remark this answers is not in the document]
Comment (vishakh) [reply zama-poc/scripts/create_bin_file.sh:4]: The test_api.sh script needs a BIN file so this serves as a little utility.

File: zama-poc/src/client.rs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+use zama_poc::zama_compute::{encrypt_genotypes_for_zama, serialize_encrypted_genotypes};
+
+// Work in progress!!
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn Error>> {
+
+    let client = Client::new();
+    let base_url = "http://localhost:6174";
+
+    let filename = "/Users/vishakh/dev/MonadicDNA/zama-poc/GFGFilteredUnphasedGenotypes23andMe.txt";
```
[the remark this answers is not in the document]
Comment (vishakh) [reply zama-poc/src/client.rs:16]: Will ask @Aeesh to do it in her client / server PR.

**Pull request merged**